### PR TITLE
Add Two of Roses

### PR DESCRIPTION
**Pull request by WarriorGallade**

### CODE REVIEW

File: data/mods/gen9ssb/abilities.ts [hunk 1]
```diff
@@ -1194,6 +1194,45 @@ export const Abilities: {[k: string]: ModdedAbilityData} = {
...
+					if (statPlus === 'accuracy' || statPlus === 'evasion') continue;
+					if (target.boosts[statPlus] < 6) {
+						stats.push(statPlus);
+					}
+				}
+				const randomStat: BoostID | undefined = stats.length ? this.sample(stats) : undefined;
+				if (randomStat) boost[randomStat] = 1;
+				this.boost(boost, target, target);
+			}
+		},
```
Comment (HisuianZoroark) [inline data/mods/gen9ssb/abilities.ts:1232]: This isn't moody

File: data/mods/gen9ssb/moves.ts [hunk 2]
```diff
@@ -2254,6 +2254,61 @@ export const Moves: {[k: string]: ModdedMoveData} = {
...
+				if (!type || type === undefined) {
+					continue;
+				} else {
+					break;
+				}
+			}
+			if (type === "Bird") type = "???";
+			if (type === "Stellar") type = pokemon.getTypes(false, true)[1];
+			move.type = type;
+		},
```
Comment (HisuianZoroark) [inline data/mods/gen9ssb/moves.ts:2286]: You can use ``let type = pokemon.getTypes()[pokemon.getTypes().length - 1];`` to get a similar effect

File: data/mods/gen9ssb/moves.ts [hunk 3]
```diff
@@ -2254,6 +2254,61 @@ export const Moves: {[k: string]: ModdedMoveData} = {
...
+					if (statPlus === 'accuracy' || statPlus === 'evasion') continue;
+					if (source.boosts[statPlus] < 6) {
+						stats.push(statPlus);
+					}
+				}
+				const randomStat: BoostID | undefined = stats.length ? this.sample(stats) : undefined;
+				if (randomStat) boost[randomStat] = 1;
+				this.boost(boost, source, source);
+			},
+		},
```
Comment (HisuianZoroark) [inline data/mods/gen9ssb/moves.ts:2302]: make this a secondary instead of a volatile

File: data/mods/gen9ssb/abilities.ts (hunk 1 above)
Comment (WarriorGallade) [reply data/mods/gen9ssb/abilities.ts:1232]: It functionally is though, boosting a random stat every turn fits the bill so i modified it to fit this spec

File: data/mods/gen9ssb/moves.ts (hunk 3 above)
Comment (WarriorGallade) [reply data/mods/gen9ssb/moves.ts:2302]: noted, will do

File: data/mods/gen9ssb/moves.ts (hunk 2 above)
Comment (WarriorGallade) [reply data/mods/gen9ssb/moves.ts:2286]: noted, will do

File: data/mods/gen9ssb/abilities.ts [hunk 4]
```diff
@@ -1194,6 +1194,47 @@ export const Abilities: {[k: string]: ModdedAbilityData} = {
 		flags: {},
 	},
 
+	// Two of Roses
+	aswesee: {
+		shortDesc: "Once per turn, when any active Pokemon has a stat boosted, this Pokemon has a 50%" +
+		" chance of copying it and a 15% chance to raise another random stat.",
+		name: "As We See",
```
Comment (KrisXV) [inline data/mods/gen9ssb/abilities.ts:1201]:
```suggestion
		name: "As We See",
		desc: "Once per turn, when any active Pokemon has a stat boosted, this Pokemon has a 50% chance of copying it and a 15% chance to raise another random stat.",
		shortDesc: "1x per turn: Stat gets boosted -> 50% chance to copy, 15% to raise another.",
```

File: data/mods/gen9ssb/abilities.ts [hunk 5]
```diff
@@ -1194,6 +1194,47 @@ export const Abilities: {[k: string]: ModdedAbilityData} = {
...
+		onFoeAfterBoost(boost, target, source, effect) { // Opportunist
+			if (this.randomChance(1, 2)) {
+				if (effect?.name === 'As We See' || effect?.name === 'Mirror Herb' || effect?.name === 'Opportunist') return;
```
Comment (KrisXV) [inline data/mods/gen9ssb/abilities.ts:1204]:
```suggestion
				if (effect && ['As We See', 'Mirror Herb', 'Opportunist'].includes(effect.name)) return;
```

File: data/mods/gen9ssb/abilities.ts [hunk 6]
```diff
@@ -1194,6 +1194,47 @@ export const Abilities: {[k: string]: ModdedAbilityData} = {
...
+				const positiveBoosts: Partial<BoostsTable> = {};
+				let i: BoostID;
+				for (i in boost) {
+					if (boost[i]! > 0) {
+						positiveBoosts[i] = boost[i];
+					}
+				}
+				if (Object.keys(positiveBoosts).length < 1) return;
+				this.boost(positiveBoosts, pokemon);
+				this.effectState.oncePerTurn = 0;
```
Comment (KrisXV) [inline data/mods/gen9ssb/abilities.ts:1215]:
```suggestion
				this.effectState.triggered = true;
```

File: data/mods/gen9ssb/abilities.ts [hunk 7]
```diff
@@ -1194,6 +1194,47 @@ export const Abilities: {[k: string]: ModdedAbilityData} = {
...
+			}
+		},
+		onResidual(target, source, effect) {
+			if (this.randomChance(15, 100) && this.effectState.oncePerTurn === 0) {
```
Comment (KrisXV) [inline data/mods/gen9ssb/abilities.ts:1219]:
```suggestion
			if (this.randomChance(15, 100) && this.effectState.triggered) {
```

File: data/mods/gen9ssb/abilities.ts [hunk 8]
```diff
@@ -1194,6 +1194,47 @@ export const Abilities: {[k: string]: ModdedAbilityData} = {
...
+					if (statPlus === 'accuracy' || statPlus === 'evasion') continue;
+					if (target.boosts[statPlus] < 6) {
+						stats.push(statPlus);
+					}
+				}
+				const randomStat: BoostID | undefined = stats.length ? this.sample(stats) : undefined;
+				if (randomStat) boost[randomStat] = 1;
+				this.boost(boost, target, target);
+			}
+			this.effectState.oncePerTurn = undefined;
```
Comment (KrisXV) [inline data/mods/gen9ssb/abilities.ts:1233]:
```suggestion
			this.effectState.triggered = false;
```

File: data/mods/gen9ssb/conditions.ts [hunk 9]
```diff
@@ -925,6 +925,71 @@ export const Conditions: {[k: string]: ModdedConditionData & {innateName?: strin
...
+					this.add('-start', target, 'typeadd', 'Fire', '[from] ability: Wonderer');
+				} else if (['sandstorm', 'deserteddunes'].includes(currentWeather) && !target.getTypes().includes('Rock')) {
+					target.addType('Rock');
+					this.add('-start', target, 'typeadd', 'Rock', '[from] ability: Wonderer');
+				} else if (['hail', 'snow'].includes(currentWeather) && !target.getTypes().includes('Ice')) {
+					target.addType('Ice');
+					this.add('-start', target, 'typeadd', 'Ice', '[from] ability: Wonderer');
+				} else {
+					// do nothing if it's not the 4 primary weathers...unless there are more?
+				}
```
Comment (KrisXV) [inline data/mods/gen9ssb/conditions.ts:964]:
```suggestion
			const currentTerrain = this.field.terrain;
			const types = target.getTypes();
			let type = 'Dark';
			if (!currentWeather && !currentTerrain && !types.includes('Dark')) {
				type = 'Dark';
			} else if (currentWeather) {
				if (['raindance', 'primordialsea'].includes(currentWeather) && !types.includes('Water')) {
					type = 'Water';
				} else if (['sunnyday', 'desolateland'].includes(currentWeather) && !types.includes('Fire')) {
					type = 'Fire';
				} else if (['sandstorm', 'deserteddunes'].includes(currentWeather) && !types.includes('Rock')) {
					type = 'Rock';
				} else if (['hail', 'snow'].includes(currentWeather) && !types.includes('Ice')) {
					type = 'Ice';
				} else {
					// do nothing if it's not the 4 primary weathers...unless there are more?
				}
```

File: data/mods/gen9ssb/conditions.ts [hunk 10]
```diff
@@ -925,6 +925,71 @@ export const Conditions: {[k: string]: ModdedConditionData & {innateName?: strin
...
+			}
+		},
+
+		onTerrainChange(target, source, sourceEffect) {
```
Comment (KrisXV) [inline data/mods/gen9ssb/conditions.ts:968]:
```suggestion
			}
			target.addType(type);
			this.add('-start', target, 'typeadd', type, '[from] ability: Wonderer');
		},
		onTerrainChange(target, source, sourceEffect) {
```

File: data/mods/gen9ssb/conditions.ts [hunk 11]
```diff
@@ -925,6 +925,71 @@ export const Conditions: {[k: string]: ModdedConditionData & {innateName?: strin
...
+					target.addType('Grass');
+					this.add('-start', target, 'typeadd', 'Grass', '[from] ability: Wonderer');
+				} else if (currentTerrain === 'mistyterrain' && !target.getTypes().includes('Fairy')) {
+					target.addType('Fairy');
+					this.add('-start', target, 'typeadd', 'Fairy', '[from] ability: Wonderer');
+				} else if (!target.getTypes().includes('Ghost')) { // custom terrains
+					target.addType('Ghost');
+					this.add('-start', target, 'typeadd', 'Ghost', '[from] ability: Wonderer');
+				}
+			}
```
Comment (KrisXV) [inline data/mods/gen9ssb/conditions.ts:990]:
```suggestion
			const currentTerrain = this.field.terrain;
			let type = 'Dark';
			const types = target.getTypes();
			if (!currentWeather && !currentTerrain && !types.includes('Dark')) {
				type = 'Dark';
			} else if (currentTerrain) {
				if (currentTerrain === 'electricterrain') {
					target.setType('Electric');
					type = '';
				} else if (currentTerrain === 'psychicterrain' && !types.includes('Psychic')) {
					type = 'Psychic';
				} else if (currentTerrain === 'grassyterrain' && !types.includes('Grass')) {
					type = 'Grass';
				} else if (currentTerrain === 'mistyterrain' && !types.includes('Fairy')) {
					type = 'Fairy';
				} else if (!types.includes('Ghost')) { // custom terrains
					type = 'Ghost';
				}
			}
			if (type) {
				target.addType(type);
				this.add('-start', target, 'typeadd', type, '[from] ability: Wonderer');
			}
```

File: data/mods/gen9ssb/moves.ts [hunk 12]
```diff
@@ -2254,6 +2254,51 @@ export const Moves: {[k: string]: ModdedMoveData} = {
 		type: "Psychic",
 	},
 
+	// Two of Roses
+	dillydally: {
+		accuracy: 90,
+		basePower: 40,
+		category: "Physical",
+		shortDesc: "Hits twice, each hit raises a random stat by 1. User's secondary type determines type.",
```
Comment (KrisXV) [inline data/mods/gen9ssb/moves.ts:2262]:
```suggestion
		shortDesc: "2 hits, +1 random stat/hit. Type = User 2nd type.",
```

File: data/mods/gen9ssb/moves.ts [hunk 13]
```diff
@@ -2254,6 +2254,51 @@ export const Moves: {[k: string]: ModdedMoveData} = {
...
+				for (statPlus in source.boosts) {
+					if (statPlus === 'accuracy' || statPlus === 'evasion') continue;
+					if (source.boosts[statPlus] < 6) {
+						stats.push(statPlus);
+					}
+				}
+				const randomStat: BoostID | undefined = stats.length ? this.sample(stats) : undefined;
+				if (randomStat) boost[randomStat] = 1;
+				this.boost(boost, source, source);
+			},
```
Comment (KrisXV) [inline data/mods/gen9ssb/moves.ts:2292]:
```suggestion
			chance: 100,
			onHit(target, source, move) {
				const stats: BoostID[] = [];
				const boost: SparseBoostsTable = {};
				let statPlus: BoostID;
				for (statPlus in source.boosts) {
					if (statPlus === 'accuracy' || statPlus === 'evasion') continue;
					if (source.boosts[statPlus] < 6) {
						stats.push(statPlus);
					}
				}
				const randomStat: BoostID | undefined = stats.length ? this.sample(stats) : undefined;
				if (randomStat) boost[randomStat] = 1;
				this.boost(boost, source, source);
			},
```

File: data/mods/gen9ssb/conditions.ts [hunk 14]
```diff
@@ -925,6 +925,71 @@ export const Conditions: {[k: string]: ModdedConditionData & {innateName?: strin
...
+				} else {
+					// do nothing if it's not the 4 primary weathers...unless there are more?
+				}
+			}
+			target.addType(type);
+			this.add('-start', target, 'typeadd', type, '[from] ability: Wonderer');
+		},
+		onTerrainChange(target, source, sourceEffect) {
+			const currentWeather = this.field.weather;
+			const currentTerrain = this.field.terrain;
```
Comment (KrisXV) [inline data/mods/gen9ssb/conditions.ts:968]:
```suggestion
			const currentWeather = this.field.getWeather().id;
			const currentTerrain = this.field.getTerrain().id;
```

File: data/mods/gen9ssb/conditions.ts [hunk 15]
```diff
@@ -925,6 +925,71 @@ export const Conditions: {[k: string]: ModdedConditionData & {innateName?: strin
...
+				if (currentTerrain === 'electricterrain') {
+					target.setType('Electric');
+					type = '';
+				} else if (currentTerrain === 'psychicterrain' && !types.includes('Psychic')) {
+					type = 'Psychic';
+				} else if (currentTerrain === 'grassyterrain' && !types.includes('Grass')) {
+					type = 'Grass';
+				} else if (currentTerrain === 'mistyterrain' && !types.includes('Fairy')) {
+					type = 'Fairy';
+				} else if (!types.includes('Ghost')) { // custom terrains
```
Comment (KrisXV) [inline data/mods/gen9ssb/conditions.ts:983]:
```suggestion
			const types = target.hasType;
			if (!currentWeather && !currentTerrain && !types('Dark')) {
				type = 'Dark';
			} else if (currentTerrain) {
				if (currentTerrain === 'electricterrain') {
					target.setType('Electric');
					type = '';
				} else if (currentTerrain === 'psychicterrain' && !types('Psychic')) {
					type = 'Psychic';
				} else if (currentTerrain === 'grassyterrain' && !types('Grass')) {
					type = 'Grass';
				} else if (currentTerrain === 'mistyterrain' && !types('Fairy')) {
					type = 'Fairy';
				} else if (!types('Ghost')) { // custom terrains
```

File: data/mods/gen9ssb/conditions.ts [hunk 16]
```diff
@@ -925,6 +925,71 @@ export const Conditions: {[k: string]: ModdedConditionData & {innateName?: strin
...
+			if (!currentWeather && !currentTerrain && !types.includes('Dark')) {
+				type = 'Dark';
+			} else if (currentWeather) {
+				if (['raindance', 'primordialsea'].includes(currentWeather) && !types.includes('Water')) {
+					type = 'Water';
+				} else if (['sunnyday', 'desolateland'].includes(currentWeather) && !types.includes('Fire')) {
+					type = 'Fire';
+				} else if (['sandstorm', 'deserteddunes'].includes(currentWeather) && !types.includes('Rock')) {
+					type = 'Rock';
+				} else if (['hail', 'snow'].includes(currentWeather) && !types.includes('Ice')) {
```
Comment (KrisXV) [inline data/mods/gen9ssb/conditions.ts:957]:
```suggestion
			const types = target.hasType;
			let type = 'Dark';
			if (!currentWeather && !currentTerrain && !types('Dark')) {
				type = 'Dark';
			} else if (currentWeather) {
				if (['raindance', 'primordialsea'].includes(currentWeather) && !types('Water')) {
					type = 'Water';
				} else if (['sunnyday', 'desolateland'].includes(currentWeather) && !types('Fire')) {
					type = 'Fire';
				} else if (['sandstorm', 'deserteddunes'].includes(currentWeather) && !types('Rock')) {
					type = 'Rock';
				} else if (['hail', 'snow'].includes(currentWeather) && !types('Ice')) {
```

**Pull request merged**